suite/tests/linux/syscall_pwait.c

```diff
@@ -93,119 +126,141 @@ main(int argc, char *argv[])
+            if (epoll_pwait(epoll_fd, &events, 24, -1, &test_set) == -1) {
+                if (errno != EINTR)
+                    perror("expected EINTR");
+            } else {
+                perror("expected interruption of syscall");
+            }
+            sigset_t post_syscall_set = {
+                0, /* Set padding to 0 so we can use memcmp */
+            };
+            pthread_sigmask(SIG_SETMASK, NULL, &post_syscall_set);
+            if (memcmp(&pre_syscall_set, &post_syscall_set, sizeof(pre_syscall_set)) !=
+                0) {
+                print("sigmask mismatch\n");
+                exit(1);
+            }
         }
     }
 
-    if (kick_off_child_signals())
-        return 0;
-
     print("Testing pselect\n");
 
-    count = 0;
-    while (count++ < 2) {
-        /* XXX i#3240: DR currently does not handle the atomicity aspect of this system
-         * call. Once it does, please include this in this test or add a new test.
-         */
-        sigset_t pre_syscall_set = {
-            0, /* Set padding to 0 so we can use memcmp */
-        };
-        sigprocmask(SIG_SETMASK, NULL, &pre_syscall_set);
-        if (pselect(0, NULL, NULL, NULL, NULL, &test_set) == -1) {
-            if (errno != EINTR)
-                perror("expected EINTR");
-        } else {
-            perror("expected interruption of syscall");
-        }
-        sigset_t post_syscall_set = {
-            0, /* Set padding to 0 so we can use memcmp */
-        };
-        sigprocmask(SIG_SETMASK, NULL, &post_syscall_set);
-        if (memcmp(&pre_syscall_set, &post_syscall_set, sizeof(pre_syscall_set)) != 0) {
-            print("sigmask mismatch");
-            exit(1);
+    for (int i = 0; i < 2; ++i) {
+        count = 0;
+        while (count++ < 2) {
+            /* XXX i#3240: DR currently does not handle the atomicity aspect of this
+             * system call. Once it does, please include this in this test or add a new
+             * test.
+             */
+            sigset_t pre_syscall_set = {
+                0, /* Set padding to 0 so we can use memcmp */
+            };
+            bool delay = i == 0 ? false : true;
+            kick_off_child_signal(count, main_thread, delay);
+            pthread_sigmask(SIG_SETMASK, NULL, &pre_syscall_set);
+            if (pselect(0, NULL, NULL, NULL, NULL, &test_set) == -1) {
+                if (errno != EINTR)
+                    perror("expected EINTR");
+            } else {
+                perror("expected interruption of syscall");
+            }
+            sigset_t post_syscall_set = {
+                0, /* Set padding to 0 so we can use memcmp */
+            };
+            pthread_sigmask(SIG_SETMASK, NULL, &post_syscall_set);
+            if (memcmp(&pre_syscall_set, &post_syscall_set, sizeof(pre_syscall_set)) !=
+                0) {
+                print("sigmask mismatch");
+                exit(1);
+            }
```

Comment:
It would be nice to share the code and avoid duplicating all of this: passing in a lambda (if this is renamed to .cpp; or a function pointer) for the p* syscall wrapper should make it relatively easy, right?

Reply:
Since this came up several times already => done.